src/theory/datatypes/datatypes_sygus.h

```diff
@@ -207,6 +207,75 @@ class SygusSymBreakNew
   };
   /** An instance of the above cache, for each anchor */
   std::map< Node, SearchCache > d_cache;
+  //-----------------------------------traversal predicates
+  /** pre/post traversal predicates for each type, variable
+   *
+   * This stores predicates (pre, post) whose semantics correspond to whether
+   * a variable has occurred by a (pre, post) traversal of a symbolic term,
+   * where index = 0 corresponds to pre, index = 1 corresponds to post. For
+   * details, see getTraversalPredicate below.
+   */
+  std::map<TypeNode, std::map<Node, Node>> d_traversal_pred[2];
+  /** traversal applications to Boolean variables
+   *
+   * This maps each application of a traversal predicate pre_x( t ) or
+   * post_x( t ) to a fresh Boolean variable.
+   */
+  std::map<Node, Node> d_traversal_bool;
+  /** get traversal predicate
+   *
+   * Get the predicates (pre, post) whose semantics correspond to whether
+   * a variable has occurred by this point in a (pre, post) traversal of a term.
+   * The type of getTraversalPredicate(tn, n, _) is tn -> Bool.
+   *
+   * For example, consider the term:
+   *   f( x_1, g( x_2, x_3 ) )
+   * and a left-to-right, depth-first traversal of this term. Let e be
+   * a variable of the same type as this term. We say that for the above term:
+   *   pre_{x_1} is false for e, e.1 and true for e.2, e.2.1, e.2.2
+   *   pre_{x_2} is false for e, e.1, e.2, e.2.1, and true for e.2.2
+   *   pre_{x_3} is false for e, e.1, e.2, e.2.1, e.2.2
+   *   post_{x_1} is true for e.1, e.2.1, e.2.2, e.2, e
+   *   post_{x_2} is false for e.1 and true for e.2.1, e.2.2, e.2, e
+   *   post_{x_3} is false for e.1, e.2.1 and true for e.2.2, e.2, e
+   *
+   * We enforce a symmetry breaking scheme for each enumerator e that is
+   * "variable-agnostic" (see argument isVarAgnostic in registerEnumerator)
+   * that ensures the variables are ordered. This scheme makes use of these
+   * predicates, described in the following:
+   *
+   * Let x_1, ..., x_m be variables that occur in the same subclass in the type
+   * of e (see TermDb::getSubclassForVar).
+   * For i = 1, ..., m:
+   *   // each variable does not occur initially in a traversal of e
+   *   ~pre_{x_i}( e ) AND
+   *   // for each subterm of e
+   *   template z.
+   *     // if this is variable x_i, then x_{i-1} must have already occurred
+   *     is-x_i( z ) => pre_{x_{i-1}}( z ) AND
+   *     for args a = 1...n
+   *       // pre-definition for each argument of this term
+   *       pre_{x_i}( z.a ) = a=0 ? pre_{x_i}( z ) : post_{x_i}( z.{a-1} ) AND
+   *     // post-definition for this term
+   *     post_{x_i}( z ) = post_{x_i}( z.n ) OR is-x_i( z )
+   *
+   * Notice that predicates pre and post should not be considered first-order
+   * predicates. For example, given the term f( x_1, x_1 ), we have that
```

Comment:
clarify that they are first-order predicates but the position of the occurrence is also an argument for its semantics. To avoid having to account for this is model construction, however, we do what you describe below.

Reply:
Here, I am saying these are Boolean variables indexed by (variable, term). This is easier since it does not require introducing a notion of position.  I will clarify this paragraph.